core/trino-spi/src/main/java/io/trino/spi/block/RowBlockEncoding.java

```diff
@@ -35,40 +37,50 @@ public String getName()
     public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceOutput, Block block)
     {
         AbstractRowBlock rowBlock = (AbstractRowBlock) block;
+        boolean isNullPresent = rowBlock.mayHaveNull();
+        int[] fieldBlockOffsets = rowBlock.getFieldBlockOffsets();
+
         int numFields = rowBlock.numFields;
 
         int positionCount = rowBlock.getPositionCount();
 
         int offsetBase = rowBlock.getOffsetBase();
-        int[] fieldBlockOffsets = rowBlock.getFieldBlockOffsets();
-        int startFieldBlockOffset = fieldBlockOffsets[offsetBase];
-        int endFieldBlockOffset = fieldBlockOffsets[offsetBase + positionCount];
+
+        int startFieldBlockOffset = fieldBlockOffsets != null ? fieldBlockOffsets[offsetBase] : offsetBase;
+        int endFieldBlockOffset = fieldBlockOffsets != null ? fieldBlockOffsets[offsetBase + positionCount] : offsetBase + positionCount;
 
         sliceOutput.appendInt(numFields);
+        sliceOutput.appendInt(positionCount);
+
         for (int i = 0; i < numFields; i++) {
             blockEncodingSerde.writeBlock(sliceOutput, rowBlock.getRawFieldBlocks()[i].getRegion(startFieldBlockOffset, endFieldBlockOffset - startFieldBlockOffset));
         }
 
-        sliceOutput.appendInt(positionCount);
-        for (int position = 0; position < positionCount + 1; position++) {
-            sliceOutput.writeInt(fieldBlockOffsets[offsetBase + position] - startFieldBlockOffset);
-        }
         EncoderUtil.encodeNullsAsBits(sliceOutput, block);
+
+        if (isNullPresent && fieldBlockOffsets != null) {
+            int[] positions = IntStream.range(0, positionCount + 1).map(i -> fieldBlockOffsets[offsetBase + i] - startFieldBlockOffset).toArray();
```

Comment:
streams are not good in tight loops. You can use `io.airlift.slice.Slices#wrappedIntArray(int[], int, int)` to point at specific offset without copying data